File: backend/tests/SasPortal.UnitTests/AdManagement/AdOperationLogCoverageMatrix.cs
namespace SasPortal.UnitTests.AdManagement;

public sealed record AdOperationLogCoverageRow(
    string OperationType,
    string LogSourceRelativePath,
    bool ExpectSuccessLog,
    bool ExpectFailureLog,
    string Notes);

public static class AdOperationLogCoverageMatrix
{
    public static IReadOnlyList<AdOperationLogCoverageRow> Rows { get; } =
    [
        new("SettingsUpdated", "backend/src/SasPortal.Persistence/Services/AdManagementSettingsService.cs", true, false, "Success-only operation log; failed saves do not write operation log."),
        new("SettingsValidated", "backend/src/SasPortal.Persistence/Services/AdManagementSettingsService.cs", true, true, "Validation summary and failure diagnostic."),
        new("AttributeMappingCreated", "backend/src/SasPortal.Persistence/Services/AdAttributeMappingService.cs", true, false, "Success-only operation log."),
        new("AttributeMappingUpdated", "backend/src/SasPortal.Persistence/Services/AdAttributeMappingService.cs", true, false, "Success-only operation log."),
        new("AttributeMappingDeleted", "backend/src/SasPortal.Persistence/Services/AdAttributeMappingService.cs", true, false, "Success-only operation log."),
        new("CreateUser", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.Create.cs", true, true, "Password excluded from request/after snapshots."),
        new("UserUpdate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.Update.cs", true, true, "Before/after user snapshots."),
        new("UserEnable", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.AccountOperations.cs", true, true, "Account status snapshots."),
        new("UserDisable", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.AccountOperations.cs", true, true, "Account status snapshots."),
        new("UserUnlock", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.AccountOperations.cs", true, true, "Lock status snapshots."),
        new("UserGroupAdd", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupMembership.cs", true, true, "Membership before/after snapshots."),
        new("UserGroupRemove", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupMembership.cs", true, true, "Membership before/after snapshots."),
        new("UserOuMove", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.MoveOu.cs", true, true, "OU move before/after snapshots."),
        new("UserManagerUpdate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ManagerUpdate.cs", true, true, "Manager before/after snapshots."),
        new("UserAccountExpirationUpdate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.AccountExpiration.cs", true, true, "Expiration before/after snapshots."),
        new("GroupCreate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupsCreate.cs", true, true, "Group after snapshot."),
        new("GroupUpdate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupsUpdate.cs", true, true, "Group before/after snapshots."),
        new("GroupDelete", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupsDelete.cs", true, true, "Group before snapshot."),
        new("GroupMemberAdd", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupsMembers.cs", true, true, "Member membership snapshots."),
        new("GroupMemberRemove", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupsMembers.cs", true, true, "Member membership snapshots."),
        new("GroupMoveOu", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.GroupsMoveOu.cs", true, true, "Group OU move snapshots."),
        new("ComputerEnable", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerAccountOperations.cs", true, true, "Computer account status snapshots."),
        new("ComputerDisable", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerAccountOperations.cs", true, true, "Computer account status snapshots."),
        new("ComputerUpdate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerUpdate.cs", true, true, "Computer before/after snapshots."),
        new("ComputerMoveOu", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerMoveOu.cs", true, true, "Computer OU move snapshots."),
        new("ComputerDelete", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerDelete.cs", true, true, "Computer delete before/after snapshots."),
        new("ComputerGroupAdd", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerGroupMembership.cs", true, true, "Computer group membership snapshots."),
        new("ComputerGroupRemove", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.ComputerGroupMembership.cs", true, true, "Computer group membership snapshots."),
        new("DeletedObjectRestore", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.DeletedObjectRestore.cs", true, true, "Deleted/restored object snapshots; credentialMode diagnostic only."),
        new("OrganizationalUnitCreate", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.OrganizationalUnitsMutations.cs", true, true, "Organizational unit after snapshot."),
        new("OrganizationalUnitRename", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.OrganizationalUnitsMutations.cs", true, true, "Organizational unit before/after rename snapshots."),
        new("OrganizationalUnitMove", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.OrganizationalUnitsMutations.cs", true, true, "Organizational unit before/after move snapshots."),
        new("OrganizationalUnitDelete", "backend/src/SasPortal.Infrastructure/Services/AdUserDirectoryService.OrganizationalUnitsMutations.cs", true, true, "Organizational unit before delete snapshot."),
    ];
}

File: frontend/src/features/ad-management/ad-operation-log-coverage-matrix.ts
import { getSnapshotRenderStrategy, type SnapshotRenderStrategy } from "./parse-ad-operation-snapshot.ts";

export type AdOperationLogCoverageRow = {
  operationType: string;
  backendConstantExists: boolean;
  frontendLabelExists: boolean;
  frontendDetailRendererExists: boolean;
  snapshotRendererExists: boolean;
  trLocaleExists: boolean;
  enLocaleExists: boolean;
  notes: string;
};

const DEDICATED_SNAPSHOT_STRATEGIES = new Set<SnapshotRenderStrategy>([
  "userUpdate",
  "userCreate",
  "groupCreate",
  "groupUpdate",
  "groupDelete",
  "computerDelete",
  "computerUpdate",
  "deletedObjectRestore",
  "accountStatus",
  "lockStatus",
  "groupMembership",
  "groupMember",
  "ouMove",
  "userManagerUpdate",
  "userAccountExpirationUpdate",
]);

/** Mirrors backend AdManagementOperationTypes public constants (inventory baseline). */
export const AD_OPERATION_LOG_COVERAGE_OPERATION_TYPES = [
  "SettingsUpdated",
  "SettingsValidated",
  "AttributeMappingCreated",
  "AttributeMappingUpdated",
  "AttributeMappingDeleted",
  "CreateUser",
  "UserUpdate",
  "UserEnable",
  "UserDisable",
  "UserUnlock",
  "UserGroupAdd",
  "UserGroupRemove",
  "UserOuMove",
  "UserManagerUpdate",
  "UserAccountExpirationUpdate",
  "GroupCreate",
  "GroupUpdate",
  "GroupDelete",
  "GroupMemberAdd",
  "GroupMemberRemove",
  "GroupMoveOu",
  "ComputerEnable",
  "ComputerDisable",
  "ComputerUpdate",
  "ComputerMoveOu",
  "ComputerDelete",
  "ComputerGroupAdd",
  "ComputerGroupRemove",
  "DeletedObjectRestore",
] as const;

export type AdOperationLogCoverageOperationType =
  (typeof AD_OPERATION_LOG_COVERAGE_OPERATION_TYPES)[number];

const COVERAGE_NOTES: Partial<Record<AdOperationLogCoverageOperationType, string>> = {
  SettingsUpdated: "Generic snapshot renderer; settings before/after snapshots.",
  SettingsValidated: "Generic snapshot renderer; validation summary only.",
  AttributeMappingCreated: "Generic snapshot renderer.",
  AttributeMappingUpdated: "Generic snapshot renderer.",
  AttributeMappingDeleted: "Generic snapshot renderer; before snapshot only on success.",
  ComputerEnable: "Uses accountStatus strategy (user-oriented sections for computer account).",
  ComputerDisable: "Uses accountStatus strategy (user-oriented sections for computer account).",
  ComputerUpdate: "Dedicated computerUpdate snapshot comparison renderer.",
  ComputerMoveOu: "ouMove strategy with computer-aware entity field grid.",
};

function hasDedicatedSnapshotRenderer(operationType: string): boolean {
  const strategy = getSnapshotRenderStrategy(operationType);
  return strategy !== "generic" && DEDICATED_SNAPSHOT_STRATEGIES.has(strategy);
}

export function buildAdOperationLogCoverageMatrix(
  trOperations: Record<string, string>,
  enOperations: Record<string, string>,
): AdOperationLogCoverageRow[] {
  return AD_OPERATION_LOG_COVERAGE_OPERATION_TYPES.map((operationType) => {
    const trLabel = trOperations[operationType];
    const enLabel = enOperations[operationType];
    const trLocaleExists = typeof trLabel === "string" && trLabel.trim().length > 0;
    const enLocaleExists = typeof enLabel === "string" && enLabel.trim().length > 0;
    const snapshotRendererExists = hasDedicatedSnapshotRenderer(operationType);

    return {
      operationType,
      backendConstantExists: true,
      frontendLabelExists: trLocaleExists && enLocaleExists,
      frontendDetailRendererExists: true,
      snapshotRendererExists,
      trLocaleExists,
      enLocaleExists,
      notes:
        COVERAGE_NOTES[operationType] ??
        (snapshotRendererExists ? "Dedicated snapshot renderer." : "Generic snapshot fallback."),
    };
  });
}

export function summarizeAdOperationLogCoverage(rows: AdOperationLogCoverageRow[]) {
  const labelCoverage = rows.filter((row) => row.frontendLabelExists).length;
  const snapshotCoverage = rows.filter((row) => row.snapshotRendererExists).length;
  const genericFallback = rows
    .filter((row) => !row.snapshotRendererExists)
    .map((row) => row.operationType);
  const localeGaps = rows
    .filter((row) => !row.trLocaleExists || !row.enLocaleExists)
    .map((row) => row.operationType);

  return {
    operationTypeCount: rows.length,
    labelCoverage,
    snapshotRendererCoverage: snapshotCoverage,
    genericFallbackOperations: genericFallback,
    localeGapOperations: localeGaps,
  };
}
